### PR TITLE
feat(evm): add shrinking to invariant testing

### PR DESCRIPTION
**Pull request by joshieDo**
## Motivation

Traces from invariant failures can really be long. Especially with a high `invariant_depth` value. This aims to reduce the sequence to its minimum representation.

## Solution
Add some basic shrinking:

```rust
/// Tries to shrink the failure case to its smallest sequence of calls.
///
/// Sets an anchor at the beginning (index=0) and tries to remove all other calls one by one,
/// until it reaches the last one. The elements which were removed and lead to a failure are
/// kept in the removal list. The removed ones that didn't lead to a failure are inserted
/// back into the sequence.
///
/// Once it reaches the end, it increments the anchor, resets the removal list and starts the
/// same process again.
///
/// Returns the smallest sequence found.
```

Is there a need for it to be optional?

### CODE REVIEW

**Review by mattsse — APPROVED** [review]
sounds reasonable, 
wdyt @onbjerg?

File: evm/src/fuzz/invariant/error.rs [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+    /// Tests that the sequence of calls successfully reverts on the error function.
+    fn fails_successfully(&self, mut executor: Executor, calls: &[&BasicTxDetails]) -> bool {
+        for (sender, (addr, bytes)) in calls {
+            executor
+                .call_raw_committing(*sender, *addr, bytes.0.clone(), 0.into())
+                .expect("bad call to evm");
+
+            // Checks the invariant.
+            if let Some(func) = &self.func {
```
Comment (gakonst) [inline evm/src/fuzz/invariant/error.rs:148]: Nit - can pull this outside of the loop and `return false` early, before iterating?

File: evm/src/fuzz/invariant/error.rs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+            let new_sequence = calls
+                .iter()
+                .enumerate()
+                .filter_map(|(index, element)| {
+                    if anchor > index || removed_calls.contains(&index) {
+                        return None
+                    }
+                    Some(element)
+                })
+                .collect::<Vec<_>>();
```
Comment (gakonst) [inline evm/src/fuzz/invariant/error.rs:191]: nit: you don't need to collect here given you'll iterate over this immutably again in `fails_successfully`

**Review by gakonst — COMMENTED** [review]
should we add a unit test which ensures that the number of calls actually is shrinked?

File: evm/src/fuzz/invariant/error.rs (hunk 1 above)
Comment (joshieDo) [reply evm/src/fuzz/invariant/error.rs]: Slightly modified it, but the goal here is to check that the invariant has, in fact, been broken after a call. 

So pulling this outside of the loop is not what we want

**Review by onbjerg — COMMENTED** [review]
the calls for invariant tests should be a proptest `ValueTree`/`Strategy`. the shrinking logic is in `ValueTree::complicate` (more calls?) and `ValueTree::simplify`. i think looking at how the value trees for a `Vec<T>` works would suffice. generally this looks ok but I'd prefer we use the "standard" approaches if possible so it's easier to migrate later/maintain as opposed to it being a totally separate thing

proptest will already rerun the failing case and call `complicate`/`simplify` to try and shrink the example, so a lot of this logic essentially duplicates that behavior

**Comment by joshieDo** [comment]
We are currently using `proptest` only for strategy/value generation on invariant tests.

Might be my inexperience with the library, but at the time of the invariant implementation I hit some limitations, which lead me to go around its standard use.

Example, for a run of depth 15, I want to be able to generate 15 values lazily (as we fill the dictionary with more in-context values), instead of getting all 15 values straightaway. Which would disregard any newly generated contracts, for example. Context:

https://github.com/foundry-rs/foundry/blob/fb9bc908d74376d657a72fa71f1675d6d050b4fc/evm/src/fuzz/invariant/executor.rs#L107-L110

from inside `test_runner.run(| | {`
https://github.com/foundry-rs/foundry/blob/fb9bc908d74376d657a72fa71f1675d6d050b4fc/evm/src/fuzz/invariant/executor.rs#L185-L193

This however, made it impossible to use the shrink functionality, since `proptest` will try to shrink with only the first generated input. That's why there's a line with `.no_shrink()` somewhere.

**Comment by onbjerg** [comment]
@joshieDo That's not entirely correct - proptest entirely discards the `ValueTree` on every test case, so the "first generated value" will be unique every time the closure we give proptest is run. In other words, the "first generated value" is in the context of a single test case (1 run), not the entire test (256 runs). All of the logic around `ValueTree`'s are only really run whenever a test case fails to try and shrink it

**Comment by onbjerg** [comment]
Talked async w @joshieDo, I don't have enough of an overview of how the primary strategy works for us to meaningfully progress on possibly using proptest for the shrinking, so you can ignore my comments on the PR - we can look at it later

**Pull request merged**